{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "train = ['population genomics of saccharomyces cerevisiae human isolates: passengers, colonizers, invaders.',\n",
    " 'estimating seed bank accumulation and dynamics in three obligateseeder proteaceae species',\n",
    " 'how and where to look for trnas in metazoan mitochondrial genomes, and what you might find when you get there',\n",
    " 'tracking global changes induced in the cd4 t cell receptor repertoire by immunization with a complex antigen using short stretches of cdr3 protein sequence.',\n",
    " 'the shrinking human protein coding complement: are there fewer than 20,000 genes?',\n",
    " 'emergence of structural and dynamical properties of ecological mutualistic networks',\n",
    " 'expertly validated models suggest responses to climate change are related to species traits: a phylogeneticallycontrolled analysis of the order lagomorpha',\n",
    " 'the emergence of the rescue effect from explicit within and betweenpatch dynamics in a metapopulation',\n",
    " 'the toxoplasma actomyoa motor complex is important but not essential for gliding motility and host cell invasion',\n",
    " 'human paternal and maternal demographic histories: insights from highresolution y chromosome and mtdna sequences']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37 97\n",
      "population genomics of saccharomyces cerevisiae human isolates: passengers, colonizers, invaders.\n",
      "True\n",
      "40 89\n",
      "estimating seed bank accumulation and dynamics in three obligateseeder proteaceae species\n",
      "True\n",
      "60 109\n",
      "how and where to look for trnas in metazoan mitochondrial genomes, and what you might find when you get there\n",
      "True\n",
      "89 156\n",
      "tracking global changes induced in the cd4 t cell receptor repertoire by immunization with a complex antigen using short stretches of cdr3 protein sequence.\n",
      "True\n",
      "42 81\n",
      "the shrinking human protein coding complement: are there fewer than 20,000 genes?\n",
      "True\n",
      "42 83\n",
      "emergence of structural and dynamical properties of ecological mutualistic networks\n",
      "True\n",
      "89 154\n",
      "expertly validated models suggest responses to climate change are related to species traits: a phylogeneticallycontrolled analysis of the order lagomorpha\n",
      "True\n",
      "47 101\n",
      "the emergence of the rescue effect from explicit within and betweenpatch dynamics in a metapopulation\n",
      "True\n",
      "64 112\n",
      "the toxoplasma actomyoa motor complex is important but not essential for gliding motility and host cell invasion\n",
      "True\n",
      "63 112\n",
      "human paternal and maternal demographic histories: insights from highresolution y chromosome and mtdna sequences\n",
      "True\n",
      "128\n"
     ]
    }
   ],
   "source": [
    "def split_len(splits):\n",
    "    total = 0\n",
    "    for s in splits:\n",
    "        total += len(s)\n",
    "    return total\n",
    "def init_vocab(splits):\n",
    "    vocab = set()\n",
    "    for s in splits:\n",
    "        for t in s:\n",
    "            vocab.add(t)\n",
    "    return list(vocab)\n",
    "\n",
    "def compute_pair_frequencies(splits: list[list[str]]):\n",
    "    counts = {}\n",
    "    for s in splits:\n",
    "        for i in range(len(s)-1):\n",
    "            k = (s[i], s[i+1])\n",
    "            if k in counts:\n",
    "                counts[k] += 1\n",
    "            else:\n",
    "                counts[k] = 1\n",
    "    return [(counts[k], k) for k in counts]\n",
    "\n",
    "\n",
    "def max_freq(freqs: list[tuple[int, str]]):\n",
    "    max_count, max_str = freqs[0]\n",
    "    for i in range(1, len(freqs)):\n",
    "        c, s = freqs[i]\n",
    "        if c > max_count:\n",
    "            max_count = c\n",
    "            max_str = s\n",
    "    return max_count, max_str\n",
    "\n",
    "def merge_pair_all(split, pair, pair_concat):\n",
    "    result = []\n",
    "    i = 0\n",
    "    while i < len(split):\n",
    "        if i < (len(split)-1) and (pair[0] == split[i] and pair[1] == split[i+1]):\n",
    "            result.append(pair_concat)\n",
    "            i+=1\n",
    "        else:\n",
    "            result.append(split[i])\n",
    "        i+=1\n",
    "    return result\n",
    "\n",
    "def split_train(train):\n",
    "    res = []\n",
    "    for t in train:\n",
    "        sub = []\n",
    "        for c in t:\n",
    "            sub.append(c)\n",
    "        res.append(sub)\n",
    "    return res\n",
    "\n",
    "class BPE:\n",
    "    def compute_bpe(self, train, n_vocab):\n",
    "        splits = split_train(train)\n",
    "        vocab = init_vocab(splits)\n",
    "        merges = []\n",
    "\n",
    "        while len(vocab) < n_vocab:\n",
    "            # count pairs\n",
    "            counts = compute_pair_frequencies(splits)\n",
    "\n",
    "            # find most frequent pair\n",
    "            freq, pair = max_freq(counts)\n",
    "\n",
    "            # add to history\n",
    "            vocab_entry = pair[0] + pair[1] # concat the string\n",
    "            merges.append((pair, vocab_entry))\n",
    "            vocab.append(vocab_entry)\n",
    "\n",
    "            for i in range(len(splits)):\n",
    "                splits[i] = merge_pair_all(splits[i], pair, vocab_entry)\n",
    "        \n",
    "        # save and return\n",
    "        self.vocab = vocab\n",
    "        self.merges = merges\n",
    "        return self\n",
    "    \n",
    "    def assert_computed(self):\n",
    "        assert self.vocab and self.merges, \"First .compute_bpe() or load_bpe()\"\n",
    "    \n",
    "    def encode(self, x: str):\n",
    "        self.assert_computed()\n",
    "        split = list(x)\n",
    "        for m in self.merges:\n",
    "            split = merge_pair_all(split, m[0], m[1])\n",
    "        return split\n",
    "\n",
    "    def save(self, filename: str):\n",
    "        self.assert_computed()\n",
    "        import pickle\n",
    "        with open(filename, \"wb\") as f:\n",
    "            pickle.dump((self.merges, self.vocab), f)\n",
    "        \n",
    "    def load(self, filename: str):\n",
    "        import pickle\n",
    "        with open(filename, \"rb\") as f:\n",
    "           self.merges, self.vocab = pickle.load(f)\n",
    "        return self\n",
    "\n",
    "bpe = BPE().compute_bpe(train, n_vocab=128)\n",
    "for t in train: \n",
    "    e = bpe.encode(t)\n",
    "    print(len(e), len(t))\n",
    "    print(''.join(e))\n",
    "    print(''.join(e) == t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
